drivers/pwm/pwm_imx.c

```diff
@@ -0,0 +1,205 @@
+
+#define PWM_PWMSR_FIFOAV_4WORDS	0x4
+
+#define PWM_PWMCR_SWR(x) (((uint32_t)(((uint32_t)(x)) \
+				<<PWM_PWMCR_SWR_SHIFT))&PWM_PWMCR_SWR_MASK)
+
+#define DEV_CFG(dev) \
+	((const struct imx_pwm_config * const)(dev)->config->config_info)
+#define DEV_DATA(dev) \
+	((struct imx_pwm_data * const)(dev)->driver_data)
+#define DEV_BASE(dev) \
+	((PWM_Type *)(DEV_CFG(dev))->base)
+
+struct imx_pwm_config {
+	PWM_Type *base;
+	u16_t prescaler;
+};
+
+struct imx_pwm_data {
+	u32_t period_cycles;
+};
+
+
+static bool imx_pwm_is_enabled(PWM_Type *base)
+{
+	return PWM_PWMCR_REG(base) & PWM_PWMCR_EN_MASK;
+}
+
+static int imx_pwm_get_cycles_per_sec(struct device *dev, u32_t pwm,
+				       u64_t *cycles)
+{
+	PWM_Type *base = DEV_BASE(dev);
+	const struct imx_pwm_config *config = DEV_CFG(dev);
+
+	*cycles = get_pwm_clock_freq(base) >> config->prescaler;
+
+	return 0;
+}
+
+static int imx_pwm_pin_set(struct device *dev, u32_t pwm,
+			    u32_t period_cycles, u32_t pulse_cycles)
+{
+	PWM_Type *base = DEV_BASE(dev);
+	const struct imx_pwm_config *config = DEV_CFG(dev);
+	struct imx_pwm_data *data = DEV_DATA(dev);
+	u8_t duty_cycle;
+	unsigned int period_ms;
+	bool enable = imx_pwm_is_enabled(base);
+	int wait_count = 0, fifoav;
+	u32_t cr, sr;
+
+
+	if ((period_cycles == 0) || (pulse_cycles > period_cycles)) {
+		SYS_LOG_ERR("Invalid combination: period_cycles=%d, "
+			    "pulse_cycles=%d", period_cycles, pulse_cycles);
+		return -EINVAL;
+	}
+
+	duty_cycle = 100 * pulse_cycles / period_cycles;
+
+	SYS_LOG_DBG("enabled=%d, pulse_cycles=%d, period_cycles=%d,"
+		    " duty_cycle=%d\n", enable, pulse_cycles, period_cycles,
+		    duty_cycle);
+
+	/*
+	 * i.MX PWMv2 has a 4-word sample FIFO.
+	 * In order to avoid FIFO overflow issue, we do software reset
+	 * to clear all sample FIFO if the controller is disabled or
+	 * wait for a full PWM cycle to get a relinquished FIFO slot
+	 * when the controller is enabled and the FIFO is fully loaded.
+	 */
+	if (enable) {
+		sr = PWM_PWMSR_REG(base);
+		fifoav = PWM_PWMSR_FIFOAV(sr);
+		if (fifoav == PWM_PWMSR_FIFOAV_4WORDS) {
+			period_ms = (get_pwm_clock_freq(base) >>
+					config->prescaler) * MSEC_PER_SEC;
+			k_sleep(K_MSEC(period_ms));
+
+			sr = PWM_PWMSR_REG(base);
+			if (fifoav == PWM_PWMSR_FIFOAV(sr))
+				SYS_LOG_WRN("there is no free FIFO slot\n");
+		}
+	} else {
+		PWM_PWMCR_REG(base) = PWM_PWMCR_SWR(1);
+		do {
+			k_sleep(1);
+			cr = PWM_PWMCR_REG(base);
+		} while ((PWM_PWMCR_SWR(cr)) &&
+			 (wait_count++ < CONFIG_PWM_PWMSWR_LOOP));
+
+		if (PWM_PWMCR_SWR(cr))
+			SYS_LOG_WRN("software reset timeout\n");
+
+	}
+
+	/*
+	 * according to imx pwm RM, the real period value should be
+	 * PERIOD value in PWMPR plus 2.
+	 */
+	if (period_cycles > 2) {
+		period_cycles -= 2;
+	} else {
+		period_cycles = 0;
+	}
+
+	PWM_PWMSAR_REG(base) = pulse_cycles;
+
+	if (data->period_cycles != period_cycles) {
+		data->period_cycles = period_cycles;
+		PWM_PWMPR_REG(base) = period_cycles;
```

Comment:
If there are still some data in the sample register FIFO, say 1-3, could this happen:
- We change the period to another value.
- Samples in the FIFO were meant to be output with the previous period value.
- But they will be output with the new period value, it may even be shorter than the sample value.

This is probably not very usual to change the period and I know the hardware does not allow to handle it better, though (unless we maintain our own fifo with both the cycle and period and set the HW FIFO low watermark to 1 and change the period in interrupt at the last moment).

Reply:
@stanislav-poboril ,

Maybe in the "if (enable)" block it can wait for a full cycle if there is still 1,2,3 or 4 words of data in the FIFO. I think this way it will only configure a new period cycle if the FIFO is empty.
What do you think?

Reply:
The best would be to try it. But then it could be perhaps too late to set the new period - there could be delays between stream of output with different cycles, which means FIFO would be emptied and the last cycle would repeat once more before changing the period.

I think the peripheral was created with the following use case in mind where period is set once and then the FIFO is filled ahead of time to keep it from underflowing (or let it underflow and repeat the same cycle until stopped or changing the cycle).

My previous comment would be probably relevant in a scenario, where the exact number of cycles has to be output (some sort of audio output?), but then I think it probably does not makes a sense to change the period duration.

Well I am not sure at the end if we should do anything about this.

Reply:
Perhaps we can leave as is (I used the same approach as the Linux kernel driver) for now and add a comment in the source code about this specific scenario that is not covered?

@MaureenHelm,
What do you think?

Reply:
I handled a similar case in pwm_mcux_ftm.c by logging a warning. You can do the same here for now.
https://github.com/zephyrproject-rtos/zephyr/blob/16ff8ca2c2aaade5b7658ccb594b9d52e1263e63/drivers/pwm/pwm_mcux_ftm.c#L61